fix(iroh): add jitter to dns retry calls

julianpryde opened the pull request:
## Description

Adds a random jitter to DNS retries called with the `stagger_call` function.

## Breaking Changes

N/A

## Notes & open questions

Further research can determine if additional timeout jitters are needed when calling the other DnsResolver public functions.

Nominal 20% of default delay time as maximum deviation from original timeout delay.  Real world and lab testing can dermine if this is sufficient.

Partially addresses #3017.

## Change checklist
<!-- Remove any that are not relevant. -->
- [x] Self-review.
- [x] Tests if relevant.

julianpryde commented on iroh-relay/src/dns.rs:
```diff
@@ -548,4 +574,35 @@ pub(crate) mod tests {
+        }
+    }
+
+    #[tokio::test]
```
Will fix in the edit

flub commented on iroh-relay/src/dns.rs:
```diff
@@ -28,6 +28,9 @@ pub const N0_DNS_NODE_ORIGIN_PROD: &str = "dns.iroh.link";
 pub const N0_DNS_NODE_ORIGIN_STAGING: &str = "staging-dns.iroh.link";
 
+/// Percent of total delay to jitter. 0.4 means +/- 20% of delay.
+const MAX_JITTER_PERCENT: u64 = 20;
```
According to the doc comment this is basically +/- 100% ?  Looking at the code I think you meant to say `20 means +/- 10% of delay`?

julianpryde replied:
Yep, I definitely forgot to change the comment. Thanks for catching that!

flub commented on iroh-relay/src/dns.rs:
```diff
@@ -548,4 +566,31 @@ pub(crate) mod tests {
         assert_eq!(result, 5)
     }
+
+    #[tokio::test]
```
This also doesn't use any async?  Just `#[test]` should do.

flub commented on iroh-relay/src/dns.rs:
```diff
@@ -548,4 +566,31 @@ pub(crate) mod tests {
+    #[traced_test]
+    fn jitter_test_nonzero_positive() {
+        let delay: u64 = 300;
+        for _ in 0..1000000{
```
How long does this test take?  I understand that you needed some confidence, but to run each time this might be a bit much.  I guess this is a usecase for proptest, but also appreciate that is a lot for this simple change.  Maybe run it with something like 10 or 100?

julianpryde replied:
I added some zeros because I was paranoid I had done the math wrong and it didn't add any noticeable delay when I ran the tests.  I can lower it though to save CI/CD compute

flub commented on iroh-relay/src/dns.rs:
```diff
@@ -548,4 +566,31 @@ pub(crate) mod tests {
+    //Sanity checks that I did the math right
+    #[test]
+    #[traced_test]
+    fn jitter_test_nonzero_positive() {
```
Probably nitpicking.  But the naming of `_negative` & `_positive` confused me.  Maybe `_lower_bound` and `_upper_bound` are better?

flub commented:
The new way of calculating is a lot easier to follow!  Thanks for these changes.

julianpryde commented:
- [x] Updated comment on MAX_JITTER_PERCENT definition to make sense
- [x] Made `jitter_test_zero` a synchronous test
- [x] Reduced number of iterations of jitter bounds tests
- [x] Renamed jitter bounds tests
- [x] ran `cargo fmt`

flub commented on iroh-relay/src/dns.rs:
```diff
@@ -28,6 +28,9 @@ pub const N0_DNS_NODE_ORIGIN_PROD: &str = "dns.iroh.link";
 pub const N0_DNS_NODE_ORIGIN_STAGING: &str = "staging-dns.iroh.link";
 
+/// Percent of total delay to jitter. 20 means +/- 20% of delay.
+const MAX_JITTER_PERCENT: u64 = 20;
```
Cool, thanks for the context.  We do retries eagerly as well to deal with slow responses, so we don't quite want to make the jittering between [0..=delay].  I'm happy to give this 20% a try and see what happens.

I suspect in any case that this isn't going to solve the entire issue.  As I said in #3017 we need to collect all callsites and think about it with a global view.
